addons/pos_loyalty/models/pos_order.py

```diff
@@ -144,6 +146,18 @@ def confirm_coupon_programs(self, coupon_data):
             'coupon_report': coupon_per_report,
         }
 
+    def check_existing_loyalty_cards(self, coupon_data):
+        coupon_key_to_modify = []
+        for coupon_id, coupon_vals in coupon_data.items():
+            partner_coupons = self.env['loyalty.card'].search(
+                [('partner_id', '=', coupon_vals.get('partner_id', False)), ('program_type', '=', 'loyalty')])
+            existing_coupons_for_program = partner_coupons.filtered(lambda c: c.program_id.id == coupon_vals['program_id'])
+            if existing_coupons_for_program:
+                coupon_vals['coupon_id'] = existing_coupons_for_program.id
```

Comment:
since some databases already has multiple loyalty cards for a partner, I think you should take it into account.